packages/commonwealth/client/scripts/views/pages/contracts/add_contract_form.tsx

```diff
@@ -0,0 +1,180 @@
+/* @jsx m */
+
+import m from 'mithril';
+import $ from 'jquery';
+
+import 'pages/create_community.scss';
+
+import app from 'state';
+import {
+  ChainBase,
+  ChainNetwork,
+  ChainType,
+  ContractType,
+} from 'common-common/src/types';
+import { isAddress } from 'web3-utils';
+
+import { notifyError } from 'controllers/app/notifications';
+import { IdRow, InputRow, SelectRow } from 'views/components/metadata_rows';
+
+import { CWButton } from 'views/components/component_kit/cw_button';
+import { CWValidationText } from 'views/components/component_kit/cw_validation_text';
+
+import { linkExistingAddressToChainOrCommunity } from 'controllers/app/login';
+
+import {
+  initChainForm,
+  defaultChainRows,
+  ethChainRows,
+} from '../create_community/chain_input_rows';
+
+import {
+  ChainFormFields,
+  ChainFormState,
+  EthChainAttrs,
+  EthFormFields,
+} from '../create_community/types';
+
+type ContractFormFields = {
+  chain_node_id: number;
+  abi: JSON;
+  contractType: ContractType;
+  decimals: number;
+  token_name: string;
+};
+
+type CreateContractForm = ChainFormFields & EthFormFields & ContractFormFields;
+
+type CreateContractState = ChainFormState & { form: CreateContractForm };
+
+export class AddContractForm implements m.ClassComponent<EthChainAttrs> {
+  private state: CreateContractState = {
+    message: '',
+    loaded: false,
+    loading: false,
+    saving: false,
+    status: undefined,
+    form: {
+      address: '',
+      chainString: 'Ethereum Mainnet',
+      // For Now we hard code the chain node id until we have a better way to distinguish between chains
+      chain_node_id: 37,
+      name: '',
+      abi: JSON.parse('[]'),
+      contractType: ContractType.ERC20,
+      nodeUrl: '',
+      symbol: '',
+      token_name: '',
+      decimals: 0,
+      ...initChainForm(),
+    },
+  };
+
+  oninit(vnode) {
+    this.state.form.nodeUrl = vnode.attrs.ethChains[1].url;
+  }
+
+  view(vnode) {
+    const validAddress = isAddress(this.state.form.address);
+
+    return (
+      <div class="CreateCommunityForm">
+        {...ethChainRows(vnode.attrs, this.state.form)}
+        <InputRow
+          title="Abi"
+          value={this.state.form.abi}
+          placeholder="Optional: Paste ABI here"
+          onChangeHandler={(value) => {
+            this.state.form.abi = value;
+            this.state.loaded = false;
+          }}
+          textarea
+        />
+        <SelectRow
+          title="Contract Type"
+          options={[
+            ContractType.ERC20,
```

Comment:
I don't really understand why we take ABIs but also only have 4 contract types available?

Reply:
this was weird, will include all contract types here